packages/analytics/amplify_analytics_pinpoint_dart/lib/src/impl/analytics_client/session_manager.dart

```diff
@@ -0,0 +1,203 @@
+      appLifecycleProvider: appLifecycleProvider,
+      onSessionStart: onSessionStart,
+      onSessionEnd: onSessionEnd,
+    );
+  }
+
+  final String _fixedEndpointId;
+  final AppLifecycleProvider? _appLifecycleProvider;
+  final OnSessionUpdated _onSessionStart;
+  final OnSessionUpdated _onSessionEnd;
+
+  _SessionCreator? _sessionCreator;
+
+  /// Get the current session
+  Session? get session => _sessionCreator?.session;
+
+  static final AmplifyLogger _logger =
+      AmplifyLogger.category(Category.analytics).createChild('SessionManager');
+
+  /// Start a new session
+  void startSession() {
+    _executeStop();
+    _executeStart();
+  }
+
+  /// Stop the current session
+  void stopSession() {
+    _executeStop();
+  }
+
+  /// Start auto tracking sessions
+  void startSessionTracking() {
+    _appLifecycleProvider?.startObserving();
+  }
+
+  /// Stop auto tracking sessions
+  void stopSessionTracking() {
+    _appLifecycleProvider?.stopObserving();
+  }
+
+  /// Get the current session state
+  SessionState getSessionState() {
+    if (_sessionCreator == null) return SessionState.inactive;
+
+    return SessionState.active;
+  }
+
+  void _executeStart() {
+    _sessionCreator = _SessionCreator.createSession(_fixedEndpointId);
+    _onSessionStart(_sessionCreator!.session);
+  }
+
+  void _executeStop() {
+    if (_sessionCreator == null) {
+      _logger.warn(
+        'Warning - stop session called without sessionBuilder initialized',
+      );
+      return;
+    }
+
+    _sessionCreator!.endSession();
+    _onSessionEnd(_sessionCreator!.session);
+
+    // Delete current session
+    _sessionCreator = null;
+  }
+}
+
+class _SessionCreator {
+  @visibleForTesting
+  const _SessionCreator({
+    required int startTimeMilliseconds,
+    required SessionBuilder sessionBuilder,
+  })  : _startTimeMilliseconds = startTimeMilliseconds,
+        _sessionBuilder = sessionBuilder;
+
+  factory _SessionCreator.createSession(String fixedEndpointId) {
+    final curTime = DateTime.now();
+
+    final sessionBuilder = SessionBuilder()
+      ..id = _generateSessionId(fixedEndpointId)
+      ..startTimestamp = curTime.toUtc().toIso8601String();
+
+    final startTimeMilliseconds = curTime.millisecondsSinceEpoch;
+
+    return _SessionCreator(
+      startTimeMilliseconds: startTimeMilliseconds,
+      sessionBuilder: sessionBuilder,
+    );
+  }
+
+  final int _startTimeMilliseconds;
+  final SessionBuilder _sessionBuilder;
+
+  static const _maxIdLength = 8;
+
+  Session get session => _sessionBuilder.build();
+
+  void endSession() {
+    final curTime = DateTime.now();
+
+    _sessionBuilder
+      ..stopTimestamp = curTime.toUtc().toIso8601String()
+      ..duration = curTime.millisecondsSinceEpoch - _startTimeMilliseconds;
+  }
+
+  static String _generateSessionId(String fixedEndpointId) {
+    var id = fixedEndpointId;
+
+    if (id.length > _maxIdLength) {
+      id = id.substring(0, _maxIdLength);
```

Comment:
Should we just be trimming the ID? Why is that appropriate? I feel like this should be an assert statement instead.

```dart
assert(id.length <= _maxIdLength, 'Max ID length exceeded');
```

Reply:
The id passed is generated via uuid.v1() which is longer than 8.  The prefix isn't as important as the remainder of the session id is a timestamp at millisecond precision.